TST: Move fully passing tests from test_interval_new to test_interval

jschendel opened the pull request:
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Some of the `test_interval_new.py` tests are fully passing, so moving them to `test_interval.py` to ensure that new changes don't break them.

Added a couple of TODO's to the new tests related to checking non-monotonic/overlapping/dupe indexes, which generally weren't covered in the new or old tests (with one noted exception).  For the most part existing new/old tests just check the non-overlapping monotonic case.

Made some additional minor modifications to `test_interval_new.py`: cleaned up a couple tests by using `parametrize`, and fixed a few instances of a misplaced parenthesis.

jschendel committed:
TST: Move fully passing tests from test_interval_new to test_interval

jschendel commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -400,16 +400,9 @@ def test_get_item(self, closed):
 
-    # To be removed, replaced by test_interval_new.py (see #16316, #16386)
+    # To be removed after overlapping/non-monotonic/duplicate cases have been
+    # added to test_get_loc_scalar below
```
The new `test_get_loc_scalar` only tests the monotonic non-overlapping case, so there are a few things in this test not fully covered.  Can add them to `test_get_loc_scalar` in this PR if desired, but didn't want to do so immediately since I didn't want to deviate too far from the existing tests.

jschendel commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -482,11 +493,18 @@ def test_slice_locs_decreasing_int64(self):
+        [(2, 4), (1, 3), (0, 2)],  # decreasing overlapping
+        [(0, 2), (0, 2), (2, 4)],  # sorted duplicates
+        [(0, 2), (2, 4), (0, 2)],  # unsorted duplicates
+        [(0, 2), (0, 2), (2, 4), (1, 3)]])  # another unsorted duplicates
```
Slight modification of the deleted test to parametrize over tuples.  Makes the test code itself cleaner.

jschendel commented on pandas/tests/indexes/interval/test_interval_new.py:
```diff
@@ -21,45 +21,25 @@ def _compare_tuple_of_numpy_array(self, result, expected):
-    def test_get_loc_interval(self, idx_side, side):
-
+    @pytest.mark.parametrize("bound", [
+        [0, 1], [1, 2], [2, 3], [3, 4], [0, 2], [2.5, 3], [-1, 4]], ids=str)
```
the entire test was within a `for` loop, pulled it out of the test by using `parametrize`

jschendel commented on pandas/tests/indexing/interval/test_interval.py:
```diff
@@ -51,9 +60,8 @@ def test_getitem_with_scalar(self):
-                             product(('increasing', 'decreasing'),
-                                     ('left', 'right', 'neither', 'both')))
+    @pytest.mark.parametrize('direction', ['increasing', 'decreasing'])
+    @pytest.mark.parametrize('closed', ['left', 'right', 'neither', 'both'])
```
minor unrelated cleanup; seems simpler just have two instances of `parametrize`

jschendel commented on pandas/tests/indexing/interval/test_interval_new.py:
```diff
@@ -214,18 +186,6 @@ def test_loc_with_overlap(self):
         with pytest.raises(KeyError):
             s.loc[1:4]
 
-    def test_non_unique(self):
```
this test looks identical to the existing one, so just deleted it entirely

jreback commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -420,8 +413,26 @@ def test_get_loc_value(self):
 
-        idx = IntervalIndex.from_arrays([0, 2], [1, 3])
-        pytest.raises(KeyError, idx.get_loc, 1.5)
+    # TODO: add overlapping/non-monotonic/duplicate index tests
```
can you add these with an xfail (or if there are a lot, leave then in _new)

jreback commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -420,8 +413,26 @@ def test_get_loc_value(self):
-        pytest.raises(KeyError, idx.get_loc, 1.5)
+    # TODO: add overlapping/non-monotonic/duplicate index tests
+    @pytest.mark.parametrize("scalar", [-0.5, 0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5])
+    def test_get_loc_scalar(self, closed, scalar):
```
can you put errors in a separate test (test_get_loc_scalar_error)

jschendel replied:
Yes, but imo doing so would be less concise add code bloat.  Could certainly do it, but want to make the downsides known first.

By separating them, you can't really parametrize over both `scalar` and `closed`, since testing endpoints will raise for some values of `closed` but pass for others, so one of them will need to be written inside a for loop of some kind.  It'd also add a bit of repetitiveness, since you'd need to encode passing keys in the passing test, and failing keys in the errors tests, whereas the current version only encodes passing with the missing values presumed to be failing.

jreback commented on pandas/tests/indexes/interval/test_interval_new.py:
```diff
@@ -21,45 +21,25 @@ def _compare_tuple_of_numpy_array(self, result, expected):
+            if bound == [0, 1]:
+                assert idx.get_loc(Interval(0, 1, closed=side)) == 0
+            elif bound == [2, 3]:
+                assert idx.get_loc(Interval(2, 3, closed=side)) == 1
```
can the error tests be pulled out to a separate test

jschendel replied:
Yes, with somewhat similar tradeoffs described above, but I don't think it'd be as bad.  This seems like it'd be more reasonable to split, but haven't fully scoped it out.

jschendel commented:
Refactored the `get_loc` tests for scalars and intervals, and added additional cases to test (overlapping, non-monotonic with dupes).  The scalar case is still fully passing.

Haven't split the errors out into separate tests yet; wanted to see if the refactoring makes things better first.  Seems like splitting the errors out would add some bloat, but can do it if that would be preferred.

jschendel committed:
refactor and add more test cases

jreback commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -400,28 +400,72 @@ def test_get_item(self, closed):
+
+        key: scalar or Interval
+
+        correct: dict
```
not a fan of this. quickly these test routines get very bespoke, and are not well known and hence repeated often enough. This is doing too much; should not catch raises. Its *ok* that it can compare an array and scalar though.

Maybe if you make a general one thats a function (and not restricted to comparisions here) would be ok.

jschendel committed:
class based testing

jschendel commented:
@jreback : what are your thoughts on the testing approach in the new file I just pushed, `test_interval_specific.py`? Haven't made any changes other than adding this file; didn't want to do too much in case I'm heading down the wrong path.

The idea is to use this approach to test behavior that depends on the underlying structure of the intervals in the index (e.g. overlapping vs. nonoverlapping monotonic), such as `get_loc`, which has specific code paths for different scenarios like this.  Behavior that is independent from the underlying interval structure (e.g. `IntervalIndex.mid`) can be left in the existing `test_interval.py` setting.

Was a little complicated to get all of this set up, but shouldn't be too bad now if someone else wanted to create a new test class variation, e.g. `TestMonotonicDecreasing`.  Not at all tied to the name `test_interval_specific.py` either; could even see this being split off into separate files for each child class, e.g. `test_interval_overlapping.py`.

jschendel commented on pandas/tests/indexes/interval/test_interval_specific.py:
```diff
@@ -0,0 +1,222 @@
+    # valid scalar tuples to test, of the from (closed abbr, scalar)
+    # (a scalar is valid if it's contained in an interval in the index)
+    scalar_tuples_valid = [
+        ('l', 0), ('l', 0.5), ('l', 2), ('l', 2.5),
```
Could remove some of these to eliminate redundancy, e.g. `('l', 0.5)` and `('l', 2.5)` are testing fundamentally the same thing.

jreback commented:
it’s well documented and pretty clear once u get pass the initial magic

@jorisvandenbossche @TomAugspurger

jschendel commented:
Closing this PR for the time being.  Not all that happy with this setup, as it seems a bit over-engineered.  Might try something similar later one once some of the `IntervalIndex` methods and API have been solidified.
